### ISSUE TITLE
issue: Bug Report: Outlet filter translations are generated but not applied to the displayed assistant message

### ISSUE THREAD

**Issue by nehogyirj** [open-webui/open-webui#30302]
### Before Submitting

- [x] I searched open and closed issues and discussions for an existing report.
- [x] I checked whether this is already fixed on the `dev` branch or latest source.
- [x] I understand that maintainers want a well-written issue before any code pull request.
- [x] I am using the latest available version of Open WebUI for my install method.
- [x] This is not a security vulnerability.

### Installation Method

Docker

### Open WebUI Version

v0.11.3

### Operating System

Debian 13

### Browser

Not browser specific problem

### Ollama Version

Not olama specific problem

### Summary

Summary

In Open WebUI v0.11.3, response/output filters appear to execute the translation request successfully, and the translation backend returns the correct translated text. However, the translated assistant response is not applied to the message displayed in the Open WebUI frontend.

This affects multiple translation filters, not only the LibreTranslate filter.

The important distinction is:

    User messages are translated successfully before being sent to the model.
    Assistant responses are sent to the translation backend.
    The translation backend returns a correct translation.
    Open WebUI still displays the original foreign-language assistant response.

Environment

    Open WebUI version: v0.11.3
    Installation: docker self-hosted
    Model: Ollama model, tested with llama3.1:8b
    Translation backend: self-hosted LibreTranslate
    LibreTranslate backend: using Argos Translate
    Filter type: Open WebUI Function Filter
    Filter API: inlet() and outlet()
    Target language: Hungarian (hu)
    Model language: English (en)

Configuration

The filter is configured with:
Source User: hu
Target User: en

Source Assistant: en
Target Assistant: hu

The filter is assigned to the model and the LibreTranslate endpoint is reachable from Open WebUI.

Not just this one, all translation filters do the same....

### Expected Behavior

Hungarian user message
        ↓
inlet filter
        ↓
English message sent to the model
        ↓
English assistant response
        ↓
outlet filter
        ↓
Hungarian assistant response displayed in Open WebUI

Example:
User input:
Szia! Hogy vagy ma?

Model input:
Hi. How are you today?

Model output:
I'm just a large language model, I don't have emotions or feelings like humans do...

Expected Open WebUI output:
Én csak egy nagy nyelvi modell vagyok, nincsenek érzelmeim vagy érzéseim...

### Actual Behavior

Actual behavior

The user message is translated correctly:
szia! Hogy vagy ma?
becomes:
Hi. How are you today?

The assistant response is sent to LibreTranslate and Argos Translate produces a correct Hungarian result.

Relevant backend log:
INFO:argostranslate.utils:('paragraphs:', ["I'm just a large language model, I don't have emotions or feelings like humans do, so I don't have a mood or feeling to express. However, I'm functioning properly and ready to assist you with any questions or tasks you have for me! How about you, how's your day going so far?"])

The translated output is then generated:
INFO:argostranslate.utils:(
    'hypotheses_to_return:',
    [(
        'Én csak egy nagy nyelvi modell vagyok, nincsenek érzelmeim vagy érzéseim, mint az embereknek, szóval nincs hangulatom, vagy érzéseim, hogy kifejezzem. Azonban, én megfelelően működik, és készen áll, hogy segítsen Önnek minden kérdést vagy feladatot nekem! És te, hogy telik a napod?',
        -20.354228496551514
    )]
)

The translation backend therefore clearly receives the English assistant response and returns a Hungarian translation.

However, Open WebUI continues to display the original English response:
I'm just a large language model, I don't have emotions or feelings like humans do, so I don't have a mood or feeling to express. However, I'm functioning properly and ready to assist you with any questions or tasks you have for me! How about you, how's your day going so far?

### Steps to Reproduce

Run Open WebUI v0.11.3.
Run a reachable LibreTranslate instance.
Install the LibreTranslate bidirectional translation filter.
Assign the filter to an Ollama model.
Configure:
source_user = hu
target_user = en
source_assistant = en
target_assistant = hu

Start a new chat.

Send:
Szia! Hogy vagy ma?

Observe that the user message is translated to English before being sent to the model.
Observe the Open WebUI/backend logs.
Confirm that the assistant response is sent to LibreTranslate.
Confirm that LibreTranslate/Argos returns a Hungarian translation.
Observe that Open WebUI still displays the original English assistant response.

Filter implementation

The filter uses the documented outlet() approach:
async def outlet(
    self,
    body: dict,
    __event_emitter__,
    **user: dict,
) -> dict:
    messages = body.get("messages", [])

    assistant_message = get_last_assistant_message(messages)

    translated_assistant_message = await self.translate(
        assistant_message,
        source_assistant,
        target_assistant,
        __event_emitter__,
        display_event_emitters,
    )

    for message in reversed(messages):
        if message["role"] == "assistant":
            message["content"] = translated_assistant_message
            break

    body["messages"] = messages
    return body

The same behavior was observed with multiple translation filters, not only this LibreTranslate implementation.
Additional observations
1. The problem is specific to the assistant/output direction

The inbound direction works:
Hungarian user message → English model input

The outbound direction fails at the Open WebUI display/application stage:
English assistant response
→ LibreTranslate
→ Hungarian translation returned
→ English response still displayed

2. The translation service is not the problem

The LibreTranslate server is reachable and responds successfully.

The Argos logs confirm:

    request received;
    English sentence segmentation performed;
    translation generated;
    Hungarian result returned.

3. The problem affects multiple translation filters

This is not limited to the LibreTranslate filter. Other translation filters that translate the assistant response in an output/outlet hook show the same behavior.
4. Streaming does not appear to be the root cause

The issue was also reproduced with streaming disabled.
5. The filter logs and translation backend logs show successful translation

The generated Hungarian text is present in the backend logs, but the Open WebUI chat message remains in English.
Suspected cause

The issue may be related to a change in the filter/output pipeline in Open WebUI v0.11.3.

Possible causes include:

    The outlet() hook receives and modifies a temporary message structure, but the frontend renders a different response object.
    The returned body["messages"] is not propagated to the frontend after the outlet filter completes.
    The final assistant message is stored or rendered from another internal representation.
    The outlet result is overwritten by the original model response after the filter completes.
    A change in the chat completion/finalization pipeline causes output filters to run after the response has already been committed to the frontend.
    The frontend receives the original assistant message from a separate websocket/event path.
    The current Open WebUI version requires a different hook, such as request(), stream(), or a message-update event, to modify the displayed output.

Diagnostic evidence

The translation result is available before the filter returns:
Translated assistant result:
Én csak egy nagy nyelvi modell vagyok, nincsenek érzelmeim...

But the visible response remains:
I'm just a large language model, I don't have emotions...

This suggests that the problem is not translation generation, language configuration, or network connectivity. It appears to be a failure to propagate the modified outlet result to the rendered assistant message.
Questions

    Is outlet() still expected to modify the assistant content displayed by the frontend in Open WebUI v0.11.3?
    Is there a known change in the filter pipeline between the version targeted by the original filter and v0.11.3?
    Should filters use request(), stream(), or another hook instead of outlet() for response translation?
    Is there a required message-update event or API for replacing the displayed assistant content?
    Does body["messages"][-1]["content"] still represent the canonical assistant message in v0.11.3?
    Is there a known issue where the original assistant response overwrites the value returned by outlet()?
    Are there differences between normal WebUI chat, Playground, and /api/chat/completions regarding outlet filter execution?

Requested fix

Please clarify or fix the output filter pipeline so that when an outlet() filter returns:
body["messages"][-1]["content"] = translated_text
return body

the modified content is consistently:

    displayed in the current chat;
    persisted in the chat history;
    used when copying the assistant response;
    used when exporting the conversation;
    used when reopening the chat.

If outlet() is no longer the correct hook for replacing the visible assistant response, the documentation should be updated with the supported method and an example implementation.
Relevant references

    Open WebUI release v0.11.2 introduced a new request() filter step for modifying the payload immediately before model calls.
    Open WebUI issue/discussion reports describe cases where outlet() modifications are not applied or are overwritten.
    Open WebUI issue #8168 reports that modifying body["messages"][-1]["content"] in outlet() can briefly appear and then revert to the original model output.
    Open WebUI discussion #8722 reports differences in filter behavior depending on the request path and notes that outlet filters may not run on certain API/Playground paths.

Short version

The translation backend works correctly. The assistant response reaches LibreTranslate, is translated successfully, and the Hungarian result is visible in the backend logs. Nevertheless, Open WebUI v0.11.3 displays the original English response. The issue affects multiple translation filters, suggesting a regression or API incompatibility in the Open WebUI outlet/output filter pipeline rather than a problem with LibreTranslate.

### Logs, Screenshots, and Config

<img width="1006" height="226" alt="Image" src="https://github.com/user-attachments/assets/a258d480-78c4-4499-aa4c-a24cb86bf4f7" />
`2026-09-21T10:35:36.144281636Z INFO:argostranslate.utils:('paragraphs:', ['Szia! Hogy vagy ma?'])

2026-09-21T10:35:36.144319788Z INFO:argostranslate.utils:('apply_packaged_translation', 'Szia! Hogy vagy ma?')

2026-09-21T10:35:36.144323566Z INFO:argostranslate.utils:('Splitting sentences using SBD Model: (hu) MiniSBDSentencizer',)

2026-09-21T10:35:36.151092622Z INFO:argostranslate.utils:('sentences', ['Szia!', 'Hogy vagy ma?'])

2026-09-21T10:35:36.160855836Z INFO:argostranslate.utils:('tokenized', [['Szia', '!'], ['Hogy', 'vagy', 'ma', '?']])

2026-09-21T10:35:36.190438359Z INFO:argostranslate.utils:('translated_batches', [TranslationResult(hypotheses=[['Hey', '.']], scores=[-1.2653441429138184], attention=[], logits=[]), TranslationResult(hypotheses=[['How', 'are', 'you', 'today', '?']], scores=[-1.0171931982040405], attention=[], logits=[])])

2026-09-21T10:35:36.190553558Z INFO:argostranslate.utils:('value_hypotheses:', [('Hey. How are you today?', -2.282537341117859)])

2026-09-21T10:35:36.190603215Z INFO:argostranslate.utils:('translated_paragraphs:', [[('Hey. How are you today?', -2.282537341117859)]])

2026-09-21T10:35:36.190658523Z INFO:argostranslate.utils:('hypotheses_to_return:', [('Hey. How are you today?', -2.282537341117859)])

2026-09-21T10:35:38.292822334Z INFO:argostranslate.utils:('paragraphs:', ['Hey! I’m doing well and ready to help. How are you today?'])

2026-09-21T10:35:38.292866669Z INFO:argostranslate.utils:('apply_packaged_translation', 'Hey! I’m doing well and ready to help. How are you today?')

2026-09-21T10:35:38.292870361Z INFO:argostranslate.utils:('Splitting sentences using SBD Model: (en) MiniSBDSentencizer',)

2026-09-21T10:35:38.293919912Z INFO:argostranslate.utils:('sentences', ['Hey!', 'I’m doing well and ready to help.', 'How are you today?'])

2026-09-21T10:35:38.294398172Z INFO:argostranslate.utils:('tokenized', [['Hey', '!'], ['I', '&apos;m', 'doing', 'well', 'and', 'ready', 'to', 'help', '.'], ['How', 'are', 'you', 'today', '?']])

2026-09-21T10:35:38.317362186Z INFO:argostranslate.utils:('translated_batches', [TranslationResult(hypotheses=[['Hé', '!']], scores=[-0.5277404189109802], attention=[], logits=[]), TranslationResult(hypotheses=[['Jól', 'vagyok', ',', 'és', 'kész', 'vagyok', 'segíteni', '.']], scores=[-3.6478590965270996], attention=[], logits=[]), TranslationResult(hypotheses=[['Hogy', 'vagy', 'ma', '?']], scores=[-1.16713285446167], attention=[], logits=[])])

2026-09-21T10:35:38.317480978Z INFO:argostranslate.utils:('value_hypotheses:', [('Hé! Jól vagyok, és kész vagyok segíteni. Hogy vagy ma?', -5.34273236989975)])

2026-09-21T10:35:38.317527609Z INFO:argostranslate.utils:('translated_paragraphs:', [[('Hé! Jól vagyok, és kész vagyok segíteni. Hogy vagy ma?', -5.34273236989975)]])

2026-09-21T10:35:38.317552299Z INFO:argostranslate.utils:('hypotheses_to_return:', [('Hé! Jól vagyok, és kész vagyok segíteni. Hogy vagy ma?', -5.34273236989975)])

2026-09-21T10:35:53.343138549Z INFO:argostranslate.utils:('paragraphs:', ["I'm just a language model, I don't have emotions or feelings like humans do, but I'm functioning properly and ready to help with any questions or tasks you have! How about you?"])

2026-09-21T10:35:53.343161863Z INFO:argostranslate.utils:('apply_packaged_translation', "I'm just a language model, I don't have emotions or feelings like humans do, but I'm functioning properly and ready to help with any questions or tasks you have! How about you?")

2026-09-21T10:35:53.343178566Z INFO:argostranslate.utils:('Splitting sentences using SBD Model: (en) MiniSBDSentencizer',)

2026-09-21T10:35:53.351363604Z INFO:argostranslate.utils:('sentences', ["I'm just a language model, I don't have emotions or feelings like humans do, but I'm functioning properly and ready to help with any questions or tasks you have!", 'How about you?'])

2026-09-21T10:35:53.351380227Z INFO:argostranslate.utils:('tokenized', [['I', '&apos;m', 'just', 'a', 'language', 'model', ',', 'I', 'don', '&apos;t', 'have', 'emotions', 'or', 'feelings', 'like', 'humans', 'do', ',', 'but', 'I', '&apos;m', 'functioning', 'properly', 'and', 'ready', 'to', 'help', 'with', 'any', 'questions', 'or', 'tasks', 'you', 'have', '!'], ['How', 'about', 'you', '?']])

2026-09-21T10:35:53.431792172Z INFO:argostranslate.utils:('translated_batches', [TranslationResult(hypotheses=[['Én', 'csak', 'egy', 'nyelvi', 'modell', 'vagyok', ',', 'nincsenek', 'érzelm@@', 'eim', 'vagy', 'érzés@@', 'eim', ',', 'mint', 'az', 'embereknek', ',', 'de', 'jól', 'működ@@', 'öm', ',', 'és', 'kész', 'vagyok', 'segíteni', 'bármilyen', 'kérdés@@', 'edben', 'vagy', 'feladat@@', 'odban', '!']], scores=[-8.27731704711914], attention=[], logits=[]), TranslationResult(hypotheses=[['És', 'te', '?']], scores=[-1.2084345817565918], attention=[], logits=[])])

2026-09-21T10:35:53.432032697Z INFO:argostranslate.utils:('value_hypotheses:', [('Én csak egy nyelvi modell vagyok, nincsenek érzelmeim vagy érzéseim, mint az embereknek, de jól működöm, és kész vagyok segíteni bármilyen kérdésedben vagy feladatodban! És te?', -9.485751628875732)])

2026-09-21T10:35:53.432056377Z INFO:argostranslate.utils:('translated_paragraphs:', [[('Én csak egy nyelvi modell vagyok, nincsenek érzelmeim vagy érzéseim, mint az embereknek, de jól működöm, és kész vagyok segíteni bármilyen kérdésedben vagy feladatodban! És te?', -9.485751628875732)]])

2026-09-21T10:35:53.432115386Z INFO:argostranslate.utils:('hypotheses_to_return:', [('Én csak egy nyelvi modell vagyok, nincsenek érzelmeim vagy érzéseim, mint az embereknek, de jól működöm, és kész vagyok segíteni bármilyen kérdésedben vagy feladatodban! És te?', -9.485751628875732)])`

**Comment by nehogyirj** [comment]
https://openwebui.com/posts/0b3a74eb-eef3-4848-908b-6ab8dcc71d98
